rfcs/plugins.md

```diff
@@ -0,0 +1,151 @@
+# Horizon plugins
+
+Plugins are intended to allow the horizon backend to be extended in a
+modular way, giving users of horizon new capabilities with a minimum
+of configuration.
+
+Plugins are intended to be flexible enough to implement things like:
+
+ - GraphQL endpoints #125)
+ - REST endpoints (#510)
+ - Custom horizon commands (#337)
+ - Custom authentication methods (#428)
+ - Server side rendering (#249)
+ - Worker processes (#368)
+ - Scheduled tasks (#311)
+
+## Things plugins can do
+ - Add new configuration options
+ - Add new `hz` commands
+ - Add new http endpoints
+ - Add new horizon request types useable by the client
+
+## Plugin interface
+ - NPM modules with the keyword 'horizon-plugin'
+ - export an object with the name 'plugin' that has one or more of the
+   following properties defined on it:
+   - `name`: a name for the plugin. Defaults to the package name
+   - `commands` for extending the `hz` command line tool [commands](#commands)
+   - `processes`: specifies subprocesses to spawn [processes](#processes)
+   - `endpoints`: specifies http endpoints that will be owned by the plugin [endpoints](#endpoints)
+   - `config`: specifies new configuration options [config](#config)
+   - `env`: specifies environment variables the plugin cares about [environment vars](#environment-vars)
+   - `requests`: specified new request types to accept over the websocket connection [requests](#requests)
+   - `activate`: Called once before the plugin is activated [activate](#activate)
+   - `deactivate`: Called once before the plugin is deactivated [deactivate](#deactivate)
+
+### Commands
+
+The `commands` plugin property must be an object with the following
+properties:
+  - `name`: a string with the name of the command to
+    implement. Defaults to an underscore-to-hyphen converted version
+    of the npm package name of the plugin
+  - `helpText`: a short description of what the command does. Mandatory.
+  - `addArguments`: a function that receives the
+    [argparse parser](http://nodeca.github.io/argparse/#HelpFormatter.prototype.addArgument)
+    for the current command, and adds any options to that parser. This
+    can include creating
+    [subparsers](http://nodeca.github.io/argparse/#ArgumentParser.prototype.addSubparsers)
+    - default is a no-op function
+  - `processConfig`: a function that receives:
```

Comment:
`addArguments` is for specifying the command line syntax, whereas `processConfig` runs after the command has been run and the options are parsed out successfully. I don't have a strong opinion, but it seems like environment variables may need more namespacing than config options or commands. Command options can be part of subcommands and the config has sections which are both like namespaces.

It'd be nice if we didn't enforce a namespacing scheme on the environment variables, or ensure that they have a specific precedence order, but maybe it's not a useful kind of flexibility.